File: packages/expo-gateflow/ios/GateFlowBridgeAdapter.swift
import Foundation
import Combine
import GateFlowKit

/**
 * Bridge adapter that connects the GateFlow iOS SDK to the Expo RN module.
 * Forwards RN method calls to the native SDK and sends SDK events back to RN.
 */
public class GateFlowBridgeAdapter {
  private let sendEvent: (String, [String: Any]) -> Void
  private var cancellables = Set<AnyCancellable>()

  init(sendEvent: @escaping (String, [String: Any]) -> Void) {
    self.sendEvent = sendEvent
  }

  func configure(apiUrl: String, apiKey: String, options: [String: Any]?) async throws {
    let sdkOptions = parseOptions(options)
    await GateFlow.shared.configure(apiUrl: apiUrl, apiKey: apiKey, options: sdkOptions)

    GateFlow.shared.$isConfigured
      .dropFirst()
      .sink { [weak self] configured in
        if configured {
          self?.sendEvent("onConfigSuccess", ["apiKey": apiKey])
        }
      }
      .store(in: &cancellables)

    GateFlow.shared.$configurationError
      .compactMap { $0 }
      .sink { [weak self] error in
        self?.sendEvent("onConfigFail", ["error": error])
      }
      .store(in: &cancellables)
  }

  func identify(userId: String, userAttrs: [String: Any]?) async throws {
    await GateFlow.shared.identify(userId)
    if let attrs = userAttrs {
      let codableAttrs = attrs.mapValues { AnyCodable($0) }
      GateFlow.shared.setUserAttributes(codableAttrs)
    }
    let attrs = GateFlow.shared.getUserAttributes()
    sendEvent("onUserIdentified", [
      "userId": userId,
      "attributes": attrs.mapValues { $0.value },
    ])
  }

  func reset() async throws {
    GateFlow.shared.reset()
    sendEvent("onUserReset", [:])
  }

  func evaluateExperiment(
    placementKey: String,
    params: [String: Any]?,
    handlerId: String?
  ) async throws -> [String: Any] {
    let result = await GateFlow.shared.evaluate(placementKey: placementKey, params: params)
    let payload = encodeResult(result, placementKey: placementKey, handlerId: handlerId)
    sendEvent("onExperimentEvaluated", payload)
    return payload
  }

  func getExperiment(key: String) async throws -> [String: Any]? {
    guard let exp = GateFlow.shared.getExperiment(key: key) else { return nil }
    return encodeExperiment(exp)
  }

  func getAllExperiments() async throws -> [[String: Any]] {
    GateFlow.shared.getAllExperiments().compactMap(encodeExperiment)
  }

  func trackEvent(event: [String: Any]) async throws {
    let gfEvent = GateFlowEvent(
      eventType: event["eventType"] as? String ?? "unknown",
      userId: event["userId"] as? String,
      experimentId: event["experimentId"] as? String,
      variantKey: event["variantKey"] as? String,
      properties: (event["properties"] as? [String: Any])?.mapValues { AnyCodable($0) }
    )
    await GateFlow.shared.track(event: gfEvent)
    sendEvent("onEventTracked", ["eventType": gfEvent.eventType])
  }

  func setUserAttributes(attrs: [String: Any]) async throws {
    let codableAttrs = attrs.mapValues { AnyCodable($0) }
    GateFlow.shared.setUserAttributes(codableAttrs)
    sendEvent("onUserAttributesUpdated", ["attributes": attrs])
  }

  func getUserAttributes() async throws -> [String: Any] {
    GateFlow.shared.getUserAttributes().mapValues { $0.value }
  }

  func preloadExperiments(placementKeys: [String]) async throws {
    await GateFlow.shared.preloadExperiments(placementKeys: placementKeys)
  }

  // MARK: - Encoding

  private func encodeResult(_ result: ExperimentResult, placementKey: String, handlerId: String?) -> [String: Any] {
    var payload: [String: Any] = ["placementKey": placementKey]
    if let handlerId = handlerId {
      payload["handlerId"] = handlerId
    }

    switch result {
    case .matched(let exp):
      payload["type"] = "matched"
      payload["experiment"] = encodeExperiment(exp)
    case .holdout(let exp):
      payload["type"] = "holdout"
      payload["experiment"] = encodeExperiment(exp)
    case .noMatch:
      payload["type"] = "noMatch"
    case .notFound:
      payload["type"] = "notFound"
    case .error(let msg):
      payload["type"] = "error"
      payload["error"] = msg
    }
    return payload
  }

  private func encodeExperiment(_ exp: Experiment) -> [String: Any] {
    [
      "id": exp.id,
      "key": exp.key,
      "layerId": exp.layerId,
      "variant": [
        "key": exp.variant.key,
        "type": exp.variant.type.rawValue,
        "params": exp.variant.params.mapValues { $0.value },
      ],
    ]
  }

  private func parseOptions(_ options: [String: Any]?) -> GateFlowOptions {
    guard let options = options else { return GateFlowOptions() }
    return GateFlowOptions(
      logLevel: parseLogLevel(options["logLevel"] as? String),
      cacheTTL: (options["cacheTTL"] as? TimeInterval) ?? 300,
      eventBatchSize: (options["eventBatchSize"] as? Int) ?? 10,
      eventFlushInterval: (options["eventFlushInterval"] as? TimeInterval) ?? 5
    )
  }

  private func parseLogLevel(_ raw: String?) -> LogLevel {
    switch raw?.lowercased() {
    case "debug": return .debug
    case "info": return .info
    case "warn", "warning": return .warn
    case "error": return .error
    default: return .info
    }
  }
}

File: packages/expo-gateflow/android/src/main/java/com/gateflow/expo/GateFlowBridgeAdapter.kt
package com.gateflow.expo

import android.content.Context
import com.gateflow.sdk.GateFlow
import com.gateflow.sdk.GateFlowOptions
import com.gateflow.sdk.delegate.GateFlowDelegate
import com.gateflow.sdk.experiment.Experiment
import com.gateflow.sdk.experiment.ExperimentResult
import com.gateflow.sdk.experiment.GateFlowEvent
import com.gateflow.sdk.logger.LogLevel

/**
 * Bridge adapter that connects the GateFlow Android SDK to the Expo RN module.
 * Forwards RN method calls to the native SDK and sends SDK events back to RN.
 */
class GateFlowBridgeAdapter(
    private val context: Context,
    private val sendEvent: (String, Map<String, Any>) -> Unit
) : GateFlowDelegate {

    suspend fun configure(apiUrl: String, apiKey: String, options: Map<String, Any>?) {
        val sdkOptions = parseOptions(options)
        GateFlow.configure(context, apiUrl, apiKey, sdkOptions)
        GateFlow.instance.delegate = this
        sendEvent("onConfigSuccess", mapOf("apiKey" to apiKey))
    }

    suspend fun identify(userId: String, userAttrs: Map<String, Any>?) {
        GateFlow.instance.identify(userId)
        if (userAttrs != null) {
            GateFlow.instance.setUserAttributes(userAttrs)
        }
        sendEvent("onUserIdentified", mapOf(
            "userId" to userId,
            "attributes" to GateFlow.instance.getUserAttributes()
        ))
    }

    suspend fun reset() {
        GateFlow.instance.reset()
        sendEvent("onUserReset", emptyMap())
    }

    suspend fun evaluateExperiment(
        placementKey: String,
        params: Map<String, Any>?,
        handlerId: String?
    ): Map<String, Any> {
        val result = GateFlow.instance.evaluate(placementKey, params)
        val payload = buildMap<String, Any> {
            put("placementKey", placementKey)
            handlerId?.let { put("handlerId", it) }
            putAll(encodeResult(result))
        }
        sendEvent("onExperimentEvaluated", payload)
        return payload
    }

    suspend fun getExperiment(key: String): Map<String, Any>? {
        val exp = GateFlow.instance.getExperiment(key) ?: return null
        return encodeExperiment(exp)
    }

    suspend fun getAllExperiments(): List<Map<String, Any>> {
        return GateFlow.instance.getAllExperiments().map(::encodeExperiment)
    }

    suspend fun trackEvent(event: Map<String, Any>) {
        val gfEvent = GateFlowEvent(
            eventType = event["eventType"] as? String ?: "unknown",
            userId = event["userId"] as? String,
            experimentId = event["experimentId"] as? String,
            variantKey = event["variantKey"] as? String,
            properties = event["properties"] as? Map<String, Any?>,
        )
        GateFlow.instance.track(gfEvent)
        sendEvent("onEventTracked", mapOf("eventType" to gfEvent.eventType))
    }

    suspend fun setUserAttributes(attrs: Map<String, Any>) {
        GateFlow.instance.setUserAttributes(attrs)
        sendEvent("onUserAttributesUpdated", mapOf("attributes" to attrs))
    }

    suspend fun getUserAttributes(): Map<String, Any> {
        @Suppress("UNCHECKED_CAST")
        return GateFlow.instance.getUserAttributes() as Map<String, Any>
    }

    suspend fun preloadExperiments(placementKeys: List<String>) {
        GateFlow.instance.preloadExperiments(placementKeys)
    }

    // MARK: - Encoding

    private fun encodeResult(result: ExperimentResult): Map<String, Any> = when (result) {
        is ExperimentResult.Matched -> mapOf(
            "type" to "matched",
            "experiment" to encodeExperiment(result.experiment),
        )
        is ExperimentResult.Holdout -> mapOf(
            "type" to "holdout",
            "experiment" to encodeExperiment(result.experiment),
        )
        is ExperimentResult.NoMatch -> mapOf("type" to "noMatch")
        is ExperimentResult.NotFound -> mapOf("type" to "notFound")
        is ExperimentResult.Error -> mapOf("type" to "error", "error" to result.message)
    }

    private fun encodeExperiment(exp: Experiment): Map<String, Any> = mapOf(
        "id" to exp.id,
        "key" to exp.key,
        "layerId" to exp.layerId,
        "variant" to mapOf(
            "key" to exp.variant.key,
            "type" to exp.variant.type.name,
            "params" to (exp.variant.params ?: emptyMap<String, Any?>()),
        ),
    )

    private fun parseOptions(options: Map<String, Any>?): GateFlowOptions {
        if (options == null) return GateFlowOptions()
        return GateFlowOptions(
            logLevel = parseLogLevel(options["logLevel"] as? String),
            cacheTTL = (options["cacheTTL"] as? Number)?.toLong() ?: 300_000L,
            eventBatchSize = (options["eventBatchSize"] as? Number)?.toInt() ?: 10,
            eventFlushInterval = (options["eventFlushInterval"] as? Number)?.toLong() ?: 5_000L,
        )
    }

    private fun parseLogLevel(raw: String?): LogLevel = when (raw?.lowercase()) {
        "debug" -> LogLevel.DEBUG
        "info" -> LogLevel.INFO
        "warn", "warning" -> LogLevel.WARN
        "error" -> LogLevel.ERROR
        else -> LogLevel.INFO
    }
}
